tools: add throttle mechanism to immutable object cache

CongMinYin opened the pull request:
tools: add throttle mechanism to immutable object cache

Implemented throttle mechanism to control promoting speed based on existing TokenBucketThrottle class.

Signed-off-by: Yin Congmin <congmin.yin@intel.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

CongMinYin commented on src/tools/immutable_object_cache/Types.cc:
```diff
@@ -90,6 +91,7 @@ void ObjectCacheReadData::encode_payload() {
   ceph::encode(read_len, payload);
   ceph::encode(pool_id, payload);
   ceph::encode(snap_id, payload);
+  ceph::encode(object_size, payload);
```
Last time you comment: "You cannot inject new fields in the middle and maintain backwards compatibility. Plus, you also need to increment the version." If I put "ceph::encode(object_size, payload);" after "ceph::encode(pool_namespace, payload);", is it OK?

dillaman replied:
Yes, it would need to go all the way at the end. You would still need to bump the encode version and in the decode you would only attempt to decode the object size field if the `version >= 2`.

CongMinYin replied:
I put the encoded at the end,  and decode it by comparing verison(Static member variable) with the current version number(16.0.0-xxxxx) now.Please make sure that my writing is correct.

CongMinYin replied:
Hi @dillaman,  I think I made a mistake. This version should come from the requester side version. However, after careful consideration, the parent cache runs on the local machine, and its communication is between  hook and daemon, both of which are on the same machine. Is it necessary to judge the version here? I don't think so. 
I am sorry that I am not familiar with these, and I will try to improve.

CongMinYin replied:
> The versioning I am talking about is for the actual messages [1] not the Ceph build sha1.
> 
> [1] https://github.com/ceph/ceph/blob/master/src/tools/immutable_object_cache/Types.cc#L20
> [2] https://github.com/ceph/ceph/blob/master/src/tools/immutable_object_cache/Types.cc#L31

In the previous patch, mykola used the version of CEPH when registering  classes: ObjectCacheRegData()
(https://github.com/ceph/ceph/blob/0ef2a830b51127987829bf400a1b3d3e8418363f/src/tools/immutable_object_cache/Types.h#L60
https://github.com/ceph/ceph/blob/0ef2a830b51127987829bf400a1b3d3e8418363f/src/tools/immutable_object_cache/CacheClient.cc#L381
https://github.com/ceph/ceph/blob/0ef2a830b51127987829bf400a1b3d3e8418363f/src/tools/immutable_object_cache/CacheClient.cc#L383). 
Before, my code also converted the string of CEPH to int and then compared. The version of CEPH seems unreasonable. So should we redefine a version here, like version=1?

dillaman replied:
That wasn't an on-the-wire format change -- it was an expectation change.

CongMinYin commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +357,71 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+      m_throttles[ROC_QOS_IOPS_THROTTLE]->get(object_num, this, &ObjectCacheStore::handle_throttle_ready, nullptr, 0);
+  }
+
+  if (got_token && (m_qos_enabled_flag & ROC_QOS_BPS_THROTTLE)) {
```
Here if blocked by IOPS,  got_token will be false, and stop promote, BPS throttle doesn't need to be done, even if it's on

CongMinYin commented:
Hi @dillaman , this throttle PR is modified based on the PR that Dehao has not completed before. (https://github.com/ceph/ceph/pull/31067)
Last time you left a few comments, I've modified, except encode: need your guidance, I left the quetion in the code section.
Thank you very much.

CongMinYin commented:
Hi @dillaman , I only tested the throttle PR in old versions. I found that the latest master can't load the immutable object cache plugin properly. I've already submit a ticket: https://tracker.ceph.com/issues/46916. Maybe you are familiar with this plugin. Can you have a look? I will also try.  After fix it, I need retest throttle in the newest master branch.

dillaman commented on src/tools/immutable_object_cache/Types.h:
```diff
@@ -85,12 +85,13 @@ class ObjectCacheReadData : public ObjectCacheRequest {
   uint64_t read_len;
   uint64_t pool_id;
   uint64_t snap_id;
+  uint64_t object_size;
```
Nit: default initialize in case it's not decoded.

CongMinYin replied:
Does default initialization mean defining a default constructor function for initialization?

dillaman replied:
You can just do this:
```suggestion
  uint64_t object_size = 0;
```

dillaman commented on src/tools/immutable_object_cache/CacheController.cc:
```diff
@@ -118,7 +118,7 @@ void CacheController::handle_request(CacheSession* session,
       int ret = m_object_cache_store->lookup_object(
         req_read_data->pool_namespace, req_read_data->pool_id,
-        req_read_data->snap_id, req_read_data->oid, return_dne_path,
+        req_read_data->snap_id, req_read_data->object_size, req_read_data->oid, return_dne_path,
```
Nit: is this line over 80 characters? If yes, please wrap it down to the next line.

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -16,6 +16,25 @@ namespace efs = std::experimental::filesystem;
 
+namespace {
+
+class SafeTimerSingleton : public SafeTimer {
```
Nit: this is using an ASIO engine so it would be better to use an ASIO timer.

CongMinYin replied:
In fact,  I don't quite understand why we need to define a singleton timer here, and we only use the lock to new TokenBucketThrottle. About ASIO timer I only know boost:: ASIO:: deadline_ Timer, I am sorry that I don't know how to do here.

dillaman replied:
I think the `TokenBucketThrottle` could be tweaked to no longer accept a `SafeTimer` and `ceph::mutex` as a parameter but instead take a `std::function` for scheduling a callback (a function that takes a lambda and returns an `Event`-like descriptor to allow the timer to be canceled). However, perhaps that's a task for a future PR.

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -35,12 +54,44 @@ ObjectCacheStore::ObjectCacheStore(CephContext *cct)
+  }
+
+  if (m_throttles[ROC_QOS_BPS_THROTTLE] != nullptr) {
+    delete m_throttles[ROC_QOS_BPS_THROTTLE];
```
Nit: why is this not covered in the case above?

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +357,71 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+  bool got_token = true;
+
+  if (m_io_throttled.load() == true) {
+    return true;
```
Shouldn't this return `false` to prevent the promotion?

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.h:
```diff
@@ -21,6 +23,11 @@ namespace immutable_obj_cache {
 typedef shared_ptr<librados::Rados> RadosRef;
 typedef shared_ptr<librados::IoCtx> IoCtxRef;
 
+enum ThrottleTargetCode {
```
Nit: does this need to be in the header?

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +357,71 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+
+  if (got_token && (m_qos_enabled_flag & ROC_QOS_IOPS_THROTTLE)) {
+    got_token =
+      m_throttles[ROC_QOS_IOPS_THROTTLE]->get(object_num, this, &ObjectCacheStore::handle_throttle_ready, nullptr, 0);
```
Nit: line length

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +357,71 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+
+  if (got_token && (m_qos_enabled_flag & ROC_QOS_BPS_THROTTLE)) {
+    got_token =
+      m_throttles[ROC_QOS_BPS_THROTTLE]->get(object_size, this, &ObjectCacheStore::handle_throttle_ready, nullptr, 0);
```
Nit: line length

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -193,7 +248,7 @@ int ObjectCacheStore::lookup_object(std::string pool_nspace,
   switch (ret) {
     case OBJ_CACHE_NONE: {
-      pret = do_promote(pool_nspace, pool_id, snap_id, object_name);
+      pret = do_promote(pool_nspace, pool_id, snap_id, object_size, object_name);
```
Nit: should this actually be put in the policy? What happens if the policy says "NONE" (i.e. please promote this) but it's never promoted since it was throttled?

CongMinYin replied:
If the policy says  OBJ_ CACHE_ NONE, and then enter "do_promote" function. Check whether the current is limited at the beginning of the function. If the current is limited, return - 1 directly. Because there is already a promote event in execution that takes up bandwidth, it will not be "never", it is just in progress.If this logic is unreasonable, it would be better to put throttle's judgment before promote instead of In the do_promote function.

dillaman replied:
I don't think the goal is to delay the actual IO operation -- but rather to delay the rate at which items are promoted to the cache. Put in another way, we don't want to rate-limit the librbd client that is performing the IO.

CongMinYin replied:
I see what you mean. This IO should be postponed instead of cancelled, otherwise it may never be promoted.I think about what to do here. It seems that I need to create a queue or another thread to process.

CongMinYin commented on src/tools/immutable_object_cache/Types.cc:
```diff
@@ -101,6 +101,31 @@ void ObjectCacheReadData::decode_payload(bufferlist::const_iterator i) {
+        tmp = 0;
+      }
+    }
+  }
```
I don't know if the number after 16.0.0 will increase with the version. I just compared the first three numbers with 16.0.0. What do you think should be done here?The next line is a normal version number:
#define CEPH_GIT_NICE_VER "16.0.0-4331-g46abf2b154"

CongMinYin replied:
ceph_version is string type, I change it to "int", extract three numbers from ceph version string and compare them with 16.0.0. Or is there a better way?

CongMinYin commented on src/tools/immutable_object_cache/Types.h:
```diff
@@ -85,18 +84,20 @@ class ObjectCacheReadData : public ObjectCacheRequest {
   uint64_t read_len;
   uint64_t pool_id;
   uint64_t snap_id;
+  uint64_t object_size = 0;
```
Just initializing this one by default , is it ok?

CongMinYin replied:
@dillaman  Just initializing this one by default , is it ok?

ceph-jenkins commented:
Can one of the admins verify this patch?

CongMinYin commented on src/tools/immutable_object_cache/Types.cc:
```diff
@@ -17,7 +17,7 @@ ObjectCacheRequest::ObjectCacheRequest(uint16_t t, uint64_t s)
 
 void ObjectCacheRequest::encode() {
-  ENCODE_START(1, 1, payload);
+  ENCODE_START(2, 1, payload);
```
@dillaman Is this the change you expect? also "DECODE_START(2, i);" and through "if (i.end())"  downward compatibility

CongMinYin commented:
Hi @dillaman , 
**version**:  see another comment.
**promote strategy**: Now store the promoting object info into a queue when CACHE_NONE, and start a thread to deal with throttle and promotion. This ensures the order of the promotion.
**Question**:  After several modifications, I didn't test it(plugin problem at before). Now I find that this PR will cause the same plugin problem as last issue. I have read your solution and still have no clue. Can you give me some suggestions?

output:
2020-09-24T01:14:40.174+0800 7f749e7fc700 -1 load failed dlopen(): "/usr/local/lib/ceph/librbd/libceph_librbd_parent_cache.so: undefined symbol: _ZNK6librbd8ImageCtx15get_object_sizeEv" or "/usr/local/lib/ceph/libceph_librbd_parent_cache.so: cannot open shared object file: No such file or directory"
2020-09-24T01:14:40.174+0800 7f749e7fc700 -1 librbd::PluginRegistry: 0x55dae0501330 init: failed to load plugin: parent_cache
2020-09-24T01:14:40.174+0800 7f749e7fc700 -1 librbd::image::OpenRequest: failed to initialize plugin registry: (38) Function not implemented
2020-09-24T01:14:40.174+0800 7f749effd700 -1 librbd::ImageState: 0x55dae0501850 failed to open image: (38) Function not implemented
rbd: error opening image testimage1: (38) Function not implemented

dillaman commented:
Have you rebased your branch to the latest master branch upstream?

CongMinYin commented:
> Have you rebased your branch to the latest master branch upstream?

yes, I rebased. No problem if removing this pr.  I'm surprised that there is no plugin related code in these modified files.

dillaman commented:
> yes, I rebased. No problem if removing this pr. I'm surprised that there is no plugin related code in these modified files.

You now invoke `ImageCtx::get_object_size` which would not be exposed to the plugin. Instead, you could pass the object-size as a constructor parameter since it's a constant.

CongMinYin commented:
> > yes, I rebased. No problem if removing this pr. I'm surprised that there is no plugin related code in these modified files.
> 
> You now invoke `ImageCtx::get_object_size` which would not be exposed to the plugin. Instead, you could pass the object-size as a constructor parameter since it's a constant.

Can plugin access m_image_ctx members but not member functions?　Can I use "1ull << m_image_ctx->order" instead of "get_object_size()"?　I test it is ok, is it right? If you agree, I will push it, otherwise we can discuss it again. see the reason :

Passing the object size through the constructor doesn't look very good.
if "get_object_size" in src/librbd/plugin/ParentCache.cc, it can't work. Before this func call, the generality of the plugin will be broken.
**init:** send_init_plugin_registry()  =>  m_image_ctx->plugin_registry->init(plugins, ctx);    =>  plugin->init(m_image_ctx, *m_plugin_api, hook_points, ctx);   =>  void ParentCache<I>::init(I* image_ctx, Api<I>& api, HookPoints* hook_points,  Context* on_finish)

**construct:** __ceph_plugin_init  =>  return plugin_registry->add( type, name, new librbd::plugin::ParentCache<librbd::ImageCtx>(cct));

dillaman commented:
> 　Can I use **"1ull << m_image_ctx->order"** instead of "get_object_size()"?

Yes, or you can probably use `m_image_ctx->layout.object_size` because the definitions of both are visible in the headers and don't rely on a symbol.

CongMinYin commented:
@dillaman Thanks your suggestion.  I have change to "m_image_ctx->layout.object_size". Could you please continue reviewing this pr. In particular, are the version and promotion strategy appropriate?

dillaman commented on src/common/options.cc:
```diff
@@ -7853,6 +7853,35 @@ static std::vector<Option> get_immutable_object_cache_options() {
     .set_default(0.1)
     .set_description("immutable object cache water mark"),
+
+    Option("immutable_object_cache_qos_schedule_tick_min", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
```
Nit: there apparently is a new `TYPE_SECS` that probably should be used for this new code (applies to `XYZ_seconds` options as well)

CongMinYin replied:
> Nit: there apparently is a new `TYPE_SECS` that probably should be used for this new code (applies to `XYZ_seconds` options as well)

But in throttle.cc, following 2 functions still use "uint64", if I change to TYPE_SECS and change val type to "secs", I still need to cast to "uint64" otherwise it will report an error.
_void TokenBucketThrottle::set_schedule_tick_min(uint64_t tick)
int TokenBucketThrottle::set_limit(uint64_t average, uint64_t burst, uint64_t burst_seconds)_

CongMinYin replied:
> You can either change `TokenBucketThrottle` to use the interval or cast it before sending it to `TokenBucketThrottle` -- but either way, we should be using the most appropriate types in the config.

I have change burst seconds type to TYPE_SECS, but "qos_schedule_tick_min" is in millisecond. There no TYPE_MILLISECS, and also can't use float in TYPE_SECS. I didn't change it.  How about "qos_schedule_tick_min".

CongMinYin replied:
> You can either change `TokenBucketThrottle` to use the interval or cast it before sending it to `TokenBucketThrottle` -- but either way, we should be using the most appropriate types in the config.

I have change burst second to TYPE_SECS, but "qos_schedule_tick_min" is in milliseconds. There is no TYPE_MILLISECS. How about it

CongMinYin replied:
I have change burst second to TYPE_SECS, but "qos_schedule_tick_min" is in milliseconds. There is no TYPE_MILLISECS. How about it

CongMinYin replied:
> You can either change `TokenBucketThrottle` to use the interval or cast it before sending it to `TokenBucketThrottle` -- but either way, we should be using the most appropriate types in the config.

I have change burst second to TYPE_SECS, but "qos_schedule_tick_min" is in milliseconds. There is no TYPE_MILLISECS. How about it

dillaman replied:
Sure, makes sense to me.

dillaman commented on src/common/options.cc:
```diff
@@ -7853,6 +7853,35 @@ static std::vector<Option> get_immutable_object_cache_options() {
+
+    Option("immutable_object_cache_qos_iops_burst_seconds", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(1)
+    .set_description("immutable object cache qos iops seconds"),
```
```suggestion
    .set_description("immutable object cache qos iops burst seconds"),
```

dillaman commented on src/tools/immutable_object_cache/Types.cc:
```diff
@@ -28,7 +28,7 @@ void ObjectCacheRequest::encode() {
 void ObjectCacheRequest::decode(bufferlist& bl) {
   auto i = bl.cbegin();
-  DECODE_START(1, i);
+  DECODE_START(2, i);
```
... and pass the `struct_v` param to `decode_payload` below

CongMinYin replied:
> ... and pass the `struct_v` param to `decode_payload` below

please check.

dillaman commented on src/tools/immutable_object_cache/Types.cc:
```diff
@@ -101,6 +103,10 @@ void ObjectCacheReadData::decode_payload(bufferlist::const_iterator i) {
   ceph::decode(snap_id, i);
   ceph::decode(oid, i);
   ceph::decode(pool_namespace, i);
+  if (i.end()) {
```
```
if (version >= 2) {
  ceph::decode(object_size, i);
}
```

dillaman commented on src/tools/immutable_object_cache/CacheController.cc:
```diff
@@ -82,6 +84,9 @@ int CacheController::run() {
       std::bind(&CacheController::handle_request, this,
                 std::placeholders::_1, std::placeholders::_2));
 
+    m_promote_thread = new std::thread([this](){
```
I'd prefer to keep this using strictly boost::asio instead of introducing a new thread.

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.h:
```diff
@@ -30,31 +32,63 @@ class ObjectCacheStore {
+  void promote_from_queue();
 
  private:
+  class Cache_Object {
```
```suggestion
  class CacheObject {
```

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -92,9 +145,9 @@ int ObjectCacheStore::init_cache() {
-                                  uint64_t pool_id, uint64_t snap_id,
-                                  std::string object_name) {
+int ObjectCacheStore::do_promote(std::string pool_nspace, uint64_t pool_id,
+                                 uint64_t snap_id, uint64_t object_size,
```
`object_size` isn't used?

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -178,25 +231,23 @@ int ObjectCacheStore::handle_promote_callback(int ret, bufferlist* read_buf,
-      }
+      Cache_Object *cache_object = new Cache_Object(pool_id, snap_id,
+        object_size, object_name, pool_nspace);
+      m_promote_queue.push(cache_object);
```
I still think this is backwards. Right now the cache policy is saying "promote this object" but if you are stuck behind QoS, you then will start to get backlogged (`m_promote_queue` will grow w/o bounds). I think the QoS logic needs to be in the policy so it can decide that it wants to promote an object because (1) it has enough space and (2) it has enough QoS available.

CongMinYin replied:
OK, so I will delete the thread code. Combined with your previous idea, I think a new state "OBJ_CACHE_THROTTLE" can be added to indicate that the cache has not been cached and can not be promoted at present because of throttle. Then implement the throttle mechanism in simplepolicy.cc. Do you think this is OK?

CongMinYin commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -193,11 +243,28 @@ int ObjectCacheStore::lookup_object(std::string pool_nspace,
+        }
+        return OBJ_CACHE_SKIP;
+      } else {
+        return ret;
```
@dillaman I deleted the thread method, still judge the throttle here. The difference is that the status is changed according to the throttle and a new state is introduced "OBJ_CACHE_THROTTLE"

CongMinYin commented:
ping @dillaman

CongMinYin commented:
@dillaman

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -193,11 +243,28 @@ int ObjectCacheStore::lookup_object(std::string pool_nspace,
+    case OBJ_CACHE_THROTTLE: {
+      if (take_token_from_throttle(object_size, 1)) {
+        m_policy->update_status(cache_file_name, OBJ_CACHE_SKIP);
+        pret = do_promote(pool_nspace, pool_id, snap_id, object_name);
```
Nit: I'm not sure we need a new state here. Just because the policy said to promote an object at time T1 (but it was then throttled) doesn't mean that the object should automatically be promoted at time T2 (if the throttle has space). Instead, I would expect the policy to return the object's state to pre time T1 (before the original `do_promote` call) so that it can be evaluated again if and only if the data is re-requested and other items haven't already moved up to higher priority within the cache policy.

CongMinYin replied:
> Nit: I'm not sure we need a new state here. Just because the policy said to promote an object at time T1 (but it was then throttled) doesn't mean that the object should automatically be promoted at time T2 (if the throttle has space). Instead, I would expect the policy to return the object's state to pre time T1 (before the original `do_promote` call) so that it can be evaluated again if and only if the data is re-requested and other items haven't already moved up to higher priority within the cache policy.

When throttling , just update status OBJ_CACHE_SKIP to OBJ_CACHE_NONE. In the next look_object(), return OBJ_CACHE_NONE directly. Here should be no need to delete entry here, _do_evict()_.

CongMinYin commented:
@dillaman see the other two comments, and do you think the configuration of  throttle should be introduced in _doc/rbd/rbd-persistent-cache.rst_?

CongMinYin commented:
@dillaman

CongMinYin commented:
@dillaman

CongMinYin commented:
@dillaman

dillaman commented:
@CongMinYin I've been pretty swamped lately, so this is on my list of things to look at next week.

CongMinYin commented:
> @CongMinYin I've been pretty swamped lately, so this is on my list of things to look at next week.

Thank you very much. I'm afraid you'll forget, so I'll Ping from time to time.

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +356,74 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+  if (!wait && (m_qos_enabled_flag & ROC_QOS_IOPS_THROTTLE)) {
+    wait =
+      m_throttles[ROC_QOS_IOPS_THROTTLE]->get(object_num, this,
+      &ObjectCacheStore::handle_throttle_ready, nullptr, 0);
```
Nit: can you clean up the indentation here (and below)?

CongMinYin replied:
> Nit: can you clean up the indentation here (and below)?

check it

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +356,74 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+    return false;
+  }
+
+  if (!wait && (m_qos_enabled_flag & ROC_QOS_IOPS_THROTTLE)) {
```
Nit: you can probably just implement this as a loop over `m_throttles` so that you can re-use the same code

CongMinYin replied:
> Nit: you can probably just implement this as a loop over `m_throttles` so that you can re-use the same code

If Iuse loops, these two sentences get more complicated because there are two sets of difference values.
(ROC_QOS_IOPS_THROTTLE, object_num)
(ROC_QOS_BPS_THROTTLE, object_size )

dillaman replied:
Fair enough

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +356,74 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+  return !wait;
+}
+
+static std::map<uint64_t, std::string> throttles_flag = {
```
```suggestion
static const std::map<uint64_t, std::string> THROTTLE_FLAGS = {
```

CongMinYin replied:
> 

If I add “const”, I will report an error when I use it to pass parameters.
error: passing ‘const std::map<long unsigned int, std::__cxx11::basic_string<char> >’ as ‘this’ argument discards qualifiers [-fpermissive]
412 |   throttle = new TokenBucketThrottle(m_cct, THROTTLE_FLAGS[flag],

dillaman replied:
Right -- you need to use something like:

```
auto throttle_flags_it = THROTTLE_FLAGS.find(flag);
ceph_assert(throttle_flags_it != THROTTLE_FLAGS.end());
throttle = new TokenBucketThrottle(m_cct, throttle_flags_it->second,
    ...);
```

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +356,74 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+  SafeTimerSingleton* safe_timer_singleton = nullptr;
+  TokenBucketThrottle* throttle = nullptr;
+  if (limit != 0) {
+    if (safe_timer_singleton == nullptr) {
```
Nit: it will also be null, right?

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +356,74 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+      throttle->set_limit(limit, 0, 1);
+    }
+  } else {
+    m_qos_enabled_flag &= ~flag;
```
Nit: seems unnecessary since you cannot dynamically change the throttles

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +356,74 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+  }
+
+  ceph_assert(m_throttles.find(flag) == m_throttles.end());
+  m_throttles.insert({flag, throttle});
```
Nit: no need to insert nullptr if throttle is not enabled, right?

CongMinYin commented:
@dillaman Please check MiLLISECS implement and func: _apply_qos_tick_and_limit_. And there's 2 concerns about your review suggestions: "loop" and "const", see another two commends.

This PR lasted a long time. If you have time, I hope it can be completed  recently. Thanks.

dillaman commented on src/common/options.h:
```diff
@@ -380,7 +391,8 @@ struct Option {
 	       || type == TYPE_UINT || type == TYPE_FLOAT
-	       || type == TYPE_SIZE || type == TYPE_SECS)))
+	       || type == TYPE_SIZE || type == TYPE_SECS
+         || type == TYPE_MILLISECS)))
```
Nit: indentation

CongMinYin commented on src/common/options.h:
```diff
@@ -377,10 +388,11 @@ struct Option {
+          && (type == TYPE_BOOL || type == TYPE_INT
+            || type == TYPE_UINT || type == TYPE_FLOAT
+            || type == TYPE_SIZE || type == TYPE_SECS
+            || type == TYPE_MILLISECS)))
```
Here are a few sentences that used the wrong tab indentation,  I changed it to space indentation.  Is this multiple indentation correct?

CongMinYin commented:
@dillaman Please check it again.

As for the microsecond unit, I found that there is another place I haven't modified:
 _src/common/options.cc:    int Option::parse_value(...) line:202_ 
But I don't know how to add MILLISECS here. 
_src/common/ceph_time.cc:  std::chrono::seconds parse_timespan(const std::string& s)_
This function parse the time with units (> seconds). Do you have any suggestions after introducing MILIIISECS units? Or let's shelve this part first,  and then do the detailed design of next PR.

dillaman commented:
@CongMinYin That `parse_timespan` is not too important for milliseconds . If we had C++20, you could just use `std::chrono::from_string`, but in this case, you can probably just use `boost::lexical_cast` to convert the string to an integer number of milliseconds.

CongMinYin commented:
@dillaman Please check  _int Option::parse_value()_ and continue reviewing.

dillaman reviewed (approved):
lgtm -- thx

dillaman commented:
jenkins test api

CongMinYin commented:
@dillaman  I see that the ceph-api test failed, so I rebased and added "millisecs" type in 
 qa/tasks/mgr/dashboard/test_cluster_configuration.py   Line:372.
Now, it passed, is it right? check it.

dillaman commented:
jenkins test dashboard

CongMinYin commented:
@dillaman  Is there any progress？

dillaman commented:
It's pending testing -- our lab is down right now so it will be next week at the earliest.

dillaman commented:
I was play-testing this today and I it doesn't seem to be working. I ran the daemon w/ the following (with some extra debug messages because I couldn't figure out why nothing was getting promoted):

```
./bin/ceph-immutable-object-cache --immutable_object_cache_path=/fast/cache --immutable_object_cache_sock=/fast/sock --immutable_object_cache_qos_iops_limit=1 -f --log-to-stderr=true --debug_immutable_obj_cache=30 2>&1  | grep -E "(promote|throttle)"
```

and what I see is that the first item kicks off a promotion and causes a throttle for the next, but `handle_throttle_ready` is never callback back until the client workload finishes. That means that in `take_token_from_throttle`, it just keeps exiting on that first conditional where it says a throttle is in-progress:

```
2020-11-30T13:04:11.313-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 do_promote: to promote object: rbd_data.102d172a10df.0000000000000010 from pool id: 2 namespace:  snapshot: 4
2020-11-30T13:04:11.313-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 promote_object: object name = rbd_data.102d172a10df.0000000000000010
2020-11-30T13:04:11.313-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 take_token_from_throttle: throttling promotion
2020-11-30T13:04:11.313-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 take_token_from_throttle: in-progress throttle
.... snip snip snip ...
2020-11-30T13:04:14.116-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 take_token_from_throttle: in-progress throttle
2020-11-30T13:04:14.116-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 take_token_from_throttle: in-progress throttle
2020-11-30T13:04:14.116-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 take_token_from_throttle: in-progress throttle
2020-11-30T13:04:14.116-0500 7fba68c4e3c0 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 take_token_from_throttle: in-progress throttle
2020-11-30T13:04:14.431-0500 7fba64887700 20 ceph::cache::ObjectCacheStore: 0x560db7b5c3a0 handle_throttle_ready: 
```

CongMinYin commented:
Hi @dillaman 
I'm sorry that my previous verification tests were not careful. I found that the problem was that daemon couldn't got token continuously(not call back func problem). That is, daemon was in the throttle state all the time. The deep-seated reason is: in the case of throttling, each IO request does not get the token, but throttle. h and throttle.cc queued up the number of the requirements.  When there is a token, use tokenbucketthrottle:: add_ tokens() automatically gets the token in throttle.cc, which results in the failure of daemon to get the token. While tokenbucketthrottle:: add_ tokens() is meaningless to get token, and also it seems meaningless to queue token requirements here.

Could you please take a look at the codes in these places and give me some suggestions? According to the code logic, this problem does occur in thropttling. I will try to see if RBD's throttle has the same problem tomorrow (I imitated RBD's throttle code)

https://github.com/ceph/ceph/blob/5c9fbc794811a1de66ee254a1f9078372839e4f3/src/common/Throttle.h#L448
https://github.com/ceph/ceph/blob/5c9fbc794811a1de66ee254a1f9078372839e4f3/src/common/Throttle.h#L425
https://github.com/ceph/ceph/blob/5c9fbc794811a1de66ee254a1f9078372839e4f3/src/common/Throttle.cc#L847

dillaman commented:
> in the case of throttling, each IO request does not get the token, but throttle. h and throttle.cc queued up the number of the requirements. When there is a token, use tokenbucketthrottle:: add_ tokens() automatically gets the token in throttle.cc, which results in the failure of daemon to get the token. While tokenbucketthrottle:: add_ tokens() is meaningless to get token, and also it seems meaningless to queue token requirements here.

If you have queued items I wouldn't expect the daemon to be able to acquire any new ones -- it needs to satisfy any outstanding requests first. Of course, the current implementation of the daemon will actually never queue up more than a single blocker at a time since otherwise it sets that boolean to say it's throttled. 

I was just seeing weird effects where it almost looked like the actual timer wasn't firing while the daemon was receiving lots of IO requests.

CongMinYin commented:
> > in the case of throttling, each IO request does not get the token, but throttle. h and throttle.cc queued up the number of the requirements. When there is a token, use tokenbucketthrottle:: add_ tokens() automatically gets the token in throttle.cc, which results in the failure of daemon to get the token. While tokenbucketthrottle:: add_ tokens() is meaningless to get token, and also it seems meaningless to queue token requirements here.
> 
> If you have queued items I wouldn't expect the daemon to be able to acquire any new ones -- it needs to satisfy any outstanding requests first. Of course, the current implementation of the daemon will actually never queue up more than a single blocker at a time since otherwise it sets that boolean to say it's throttled.
> 
> I was just seeing weird effects where it almost looked like the actual timer wasn't firing while the daemon was receiving lots of IO requests.

I don't queued items in daemon, but throttle queued every I/O request token.

CongMinYin commented:
@dillaman  This PR closed because of my force-push.  I cann't reopen it. What should I do?

I have fixed problem.
Tokenbucketthrottle is ordered. Once there is a token, the callback function will be executed to continue IO.   When the timer add the tokens, request in throttle list gets the token and call back to expect continuing the IO request at that time. However, for the cache, the promotion at that time has been abandoned, Therefore, I collected the token in the callback function and used it for the present promotion.

tchaikov commented:
@CongMinYin, you could

1. stash your latest work, 
2. `git reset --hard 6d4f0b3659785bbac7b94883a7ea32417085d1c5`
3. force push again
4. restore your latest work

CongMinYin commented:
> @CongMinYin, you could
> 
> 1. stash your latest work,
> 2. `git reset --hard 6d4f0b3659785bbac7b94883a7ea32417085d1c5`
> 3. force push again
> 4. restore your latest work

Thanks

CongMinYin commented:
jenkins retest this please

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.h:
```diff
@@ -48,13 +55,27 @@ class ObjectCacheStore {
+  uint64_t m_qos_enabled_flag{0};
+  std::map<uint64_t, TokenBucketThrottle*> m_throttles;
+  std::atomic<bool> m_io_throttled{false};
+  ceph::mutex m_throttle_lock;
```
Lock isn't initialized:

```
error: no matching function for call to ‘ceph::mutex_debug_detail::mutex_debug_impl<false>::mutex_debug_impl()’
```

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.h:
```diff
@@ -48,13 +55,27 @@ class ObjectCacheStore {
+  // throttle mechanism
+  uint64_t m_qos_enabled_flag{0};
+  std::map<uint64_t, TokenBucketThrottle*> m_throttles;
+  std::atomic<bool> m_io_throttled{false};
```
Nit: no need for an atomic type anymore

dillaman commented on src/tools/immutable_object_cache/ObjectCacheStore.cc:
```diff
@@ -302,5 +361,95 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
+  bool wait = false;
+  if (!wait && (m_qos_enabled_flag & ROC_QOS_IOPS_THROTTLE)) {
+    std::lock_guard lock(m_throttle_lock);
+    if (object_num < m_iops_tokens) {
```
`m_iops_tokens` is initialized to zero so this will always fail the test on the first invokation, which then hits line 391 which subtracts 1 from 0 (to get a very large positive number).

CongMinYin commented:
jenkins retest this please

CongMinYin commented:
@dillaman  Corrected the fatal mistake,  init mutex , please check.

I'm very sorry that I didn't find any problems every time I did the verification test.
